pkg/util/kubernetes/apiserver/services.go

```diff
@@ -0,0 +1,73 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+// +build kubeapiserver
+
+package apiserver
+
+import (
+	"errors"
+
+	log "github.com/cihub/seelog"
+
+	"github.com/DataDog/datadog-agent/pkg/util/cache"
+	"github.com/ericchiang/k8s/api/v1"
+)
+
+// mapServices maps each pod (endpoint) to the services connected to it.
+// It is on a per node basis to avoid mixing up the services pods are actually connected to if all pods of different nodes share a similar subnet, therefore sharing a similar IP.
+func (smb *ServiceMapperBundle) mapServices(nodeName string, pods v1.PodList, endpointList v1.EndpointsList) error {
+	smb.m.Lock()
+	defer smb.m.Unlock()
+	ipToEndpoints := make(map[string][]string) // maps the IP address from an endpoint (pod) to associated services ex: "10.10.1.1" : ["service1","service2"]
+	podToIp := make(map[string]string)         // maps the pods of the currently evaluated node to their IP.
+
+	if pods.Items == nil {
+		err := errors.New("empty podlist received")
+		return err
+	}
+
+	for _, pod := range pods.Items {
+		if *pod.Status.PodIP != "" {
+			podToIp[*pod.Metadata.Name] = *pod.Status.PodIP
+		}
+	}
+	for _, svc := range endpointList.Items {
+		for _, endpointsSubsets := range svc.Subsets {
+			if endpointsSubsets.Addresses == nil {
+				log.Tracef("A subset of endpoints from %s could not be evaluated", *svc.Metadata.Name)
+				continue
+			}
+			for _, edpt := range endpointsSubsets.Addresses {
+				if edpt == nil {
```

Comment:
Is it possible to have a `nil` element in the iterable object `endpointsSubsets.Addresses` ?

Reply:
Not sure, the strategy here is to avoid evaluating a nil pointer. Do you think it should be removed ?